specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1096 @@
+
+Whenever a new bottom-up message is triggered in a subnet, its `SCA`:
+- Burns the amount of native tokens included in the `value` of the message in the subnet.
+- Checks the checkpoint being populated in the current checkpoint window and checks if it already has a `CrossMsgMeta` with the same destination of the message.
+    - If the `CrossMsgMeta` doesn't exist in the checkpoint, the SCA creates a new `CrossMsgs` appending the cross-net message in the `CrossMsgsRegistry` of the SCA, and includes in the checkpoint a new `CrossMsgMeta` for the destination including the `CID` of the `CrossMsgs` stored in the registry.
+    - If the `CrossMsgMeta` for the destination exists in the checkpoint, the SCA gets from the `CrossMsgsRegistry` the current `CID` included in the `CrossMsgMeta`, and it appends the newly created cross-net message to `Msgs`. `SCA` then updates with the new `CID` (after appending the message) to the `CrossMsgMeta` for the checkpoint, deletes the outdated `CrossMsgs` from the registry, and includes the updated one.
+    - In these updates, also the total amount of native tokens included in the messages of the `CrossMsgMeta` is updated in the `Value` field.
+- Finally, when the signing window for the checkpoint closes, the checkpoint is propagated including a link to the cross-net message in the `CrossMsgMeta` of the checkpoint. 
+
+> Bottom-up messages and checkpoint propagation data structures
+```go
+
+// Metadata pointing to the list of messages being
+// propagated in the checkpoint.
+type CrossMsgMeta struct {
+    // Source of the cross-net messages included in
+    // this metadata package
+    From    address.SubnetID
+    // Destination of the cross-net messages included.
+    To      address.SubnetID
+    // CID of the aggregation of all the messages being
+    // propagated. This CID is used to uniquely identify
+    // this package of messages.
+    MsgsCid cid.Cid
+    // Nonce of the crossMsgMeta. It is used for partial
+    // ordering and to prevent replay attacks.
+    Nonce   uint64
+    // Aggregation of all the native tokens being transacted
+    // in the included messages.
+    Value   abi.TokenAmount
+}
+
+// CrossMsgs is the data structure used to persist in the
+// `CrossMsgsRegistry` the `Msgs` and `CrossMsgMeta` 
+// propagated in checkpoints
+type CrossMsgs struct {
+    // Raw msgs from the subnet
+    Msgs  []Message   
+    // Metas propagated from child subnets
+    Metas []CrossMsgMeta 
+}
+
+// MetaTag is a convenient struct
+// used to compute the Cid of the MsgMeta
+type MetaTag struct {
+    MsgsCid  cid.Cid
+    MetasCid cid.Cid
```

Comment:
It would be good to have at least a comment to what kind of `Cid` these are, and in all other raw Cid cases as well.
I tried filling them in in my PRs but I find myself going in circles in the code to figure out if I did the right thing.

Reply:
Fixed. I typed every CID in the spec as to complement your PRs. Thanks. (I will push the PR with the update soon)